### PR TITLE
Fix hotkeys for Linux

### PR DESCRIPTION
**Pull request by randomstuff**
* Make the keybindings work when one key of the expression is not available;
* Fix default hotkey bindings for Linux ;
* Avoid conflict with global shortcuts.

### CODE REVIEW

**Comment by skidau** [comment]
Nice, I think this will fix a long and annoying bug that I was encountering with the Expression parser.

File: Source/Core/InputCommon/ControllerInterface/ExpressionParser.cpp [hunk 1]
```diff
@@ -219,6 +219,33 @@ class ExpressionNode
...
+	}
+
+	virtual int CountNumControls() override
+	{
+		return 0;
+	}
+
+	virtual operator std::string() override
+	{
+		return "`" + (std::string)name + "`";
```
Comment (lioncash) [inline Source/Core/InputCommon/ControllerInterface/ExpressionParser.cpp]: The cast here is redundant considering name is a string

File: Source/Core/InputCommon/ControllerInterface/ExpressionParser.cpp [hunk 2]
```diff
@@ -219,6 +219,33 @@ class ExpressionNode
 	virtual operator std::string() { return ""; }
 };
 
+class DummyExpression : public ExpressionNode
+{
+public:
+	std::string name;
+
+	DummyExpression(std::string const& name_) : name(name_) {}
```
Comment (lioncash) [inline Source/Core/InputCommon/ControllerInterface/ExpressionParser.cpp]: `const std::string&`

File: Source/Core/InputCommon/ControllerInterface/ExpressionParser.cpp [hunk 3]
```diff
@@ -219,6 +219,33 @@ class ExpressionNode
...
+
+	virtual ControlState GetValue() override
```
Comment (lioncash) [inline Source/Core/InputCommon/ControllerInterface/ExpressionParser.cpp]: You can remove the `virtual` specifiers off the overridden functions.

File: Source/Core/InputCommon/ControllerInterface/ExpressionParser.cpp [hunk 4]
```diff
@@ -415,8 +442,10 @@ class Parser
 		case TOK_CONTROL:
 			{
 				Device::Control *control = finder.FindControl(tok.qualifier);
-				if (control == nullptr)
-					return EXPRESSION_PARSE_NO_DEVICE;
+				if (control == nullptr) {
```
Comment (lioncash) [inline Source/Core/InputCommon/ControllerInterface/ExpressionParser.cpp]: Brace on next line.

File: Source/Core/InputCommon/ControllerInterface/ExpressionParser.cpp [hunk 5]
```diff
@@ -219,6 +219,33 @@ class ExpressionNode
...
+	std::string name;
+
+	DummyExpression(std::string const& name_) : name(name_) {}
+
+	virtual ControlState GetValue() override
+	{
+		return 0.0;
+	}
+
+	virtual void SetValue(ControlState value_) override
```
Comment (lioncash) [inline Source/Core/InputCommon/ControllerInterface/ExpressionParser.cpp]: `value`

**Comment by skidau** [comment]
@dolphin-emu-bot rebuild

**Comment by skidau** [comment]
@dolphin-emu-bot rebuild

**Comment by skidau** [comment]
The "Make the keybindings work when one key of the expression is not available;" has fixed the annoying bug where my controls wouldn't work until I plugged in a controller.  Many thanks.

**Comment by randomstuff** [comment]
@skidau: I guess this was a case where a keybinding could be done either with a controller or a keyboard, `(SomeKeyboardControl | SomeGamepadControl)`?

**Comment by skidau** [comment]
Correct

**Comment by randomstuff** [comment]
OK, great.

**Pull request merged**